### PR TITLE
WIP: install package refactor

### PR DESCRIPTION
**Pull request by asalkeld**
closes: https://github.com/Azure/ARO-RP/issues/106

I am not completely happy with this, but just to get some ideas from others.

### CODE REVIEW

File: pkg/api/openshiftcluster.go [hunk 1]
```diff
@@ -201,7 +201,6 @@ type InstallPhase int
 
 // InstallPhase constants
 const (
-	InstallPhaseDeployStorage InstallPhase = iota
-	InstallPhaseDeployResources
+	InstallPhaseDeploy InstallPhase = iota
```
Comment (mjudeikis) [inline pkg/api/openshiftcluster.go]: I dont know. I still reluctant on this. Install phase in my head should encapsulate only installer work, and we should have separate phase for "post install" things. And remove bootstrap is last. Im fine merging it, but I think this is not the solution to condense long running operations to even more long running operations.

Comment (jim-minter) [reply pkg/api/openshiftcluster.go]: We have to minimise the number of overt install phases otherwise RP upgrade becomes extremely challenging.

**Comment by jim-minter** [comment]
> I am not completely happy with this, but just to get some ideas from others.

Yeah.  I think this is too much change in one go, and perhaps it loses sight of the aim.

How about starting by focusing on the 0-, 1-, 2- files: each has one uber-function in it which has multiple independent steps, each of which are meant to be restartable (goal based, whether acting or waiting for a state).

We know we can't expose each of those steps as a unique phase in the DB, but can we standardise them into separate files with very similar or identical function signatures such that at least theoretically each phase could become table driven?  (each phase could get to become a table of function pointers, each function is testably restartable, the "uber" function walks the table calling the functions one by one?)

It's just an idea, perhaps sketch it out on the shortest file (2- ?) and see what it looks like?

Maybe try that and see what comes of it before modifying the template or graph pieces?

**Comment by asalkeld** [comment]
> > I am not completely happy with this, but just to get some ideas from others.
> 
> Yeah. I think this is too much change in one go, and perhaps it loses sight of the aim.
> 
> How about starting by focusing on the 0-, 1-, 2- files: each has one uber-function in it which has multiple independent steps, each of which are meant to be restartable (goal based, whether acting or waiting for a state).
> 
> We know we can't expose each of those steps as a unique phase in the DB, but can we standardise them into separate files with very similar or identical function signatures such that at least theoretically each phase could become table driven? (each phase could get to become a table of function pointers, each function is testably restartable, the "uber" function walks the table calling the functions one by one?)
> 
> It's just an idea, perhaps sketch it out on the shortest file (2- ?) and see what it looks like?
> 
> Maybe try that and see what comes of it before modifying the template or graph pieces?

Great idea here it is, I like it: https://github.com/Azure/ARO-RP/pull/163

**Comment by asalkeld** [comment]
closing this in favour of #163

**Pull request closed without merging**